tests/python_package_test/test_dask.py

```diff
@@ -826,3 +826,29 @@ def test_dask_classes_and_sklearn_equivalents_have_identical_constructors_except
     assert dask_spec.defaults[:-1] == sklearn_spec.defaults
     assert dask_spec.args[-1] == 'client'
     assert dask_spec.defaults[-1] is None
+
+
+@pytest.mark.parametrize(
+    "methods",
+    [
+        (lgb.DaskLGBMClassifier.fit, lgb.LGBMClassifier.fit),
+        (lgb.DaskLGBMClassifier.predict, lgb.LGBMClassifier.predict),
+
+        (lgb.DaskLGBMRegressor.fit, lgb.LGBMRegressor.fit),
+        (lgb.DaskLGBMRegressor.predict, lgb.LGBMRegressor.predict),
+
+        (lgb.DaskLGBMRanker.fit, lgb.LGBMRanker.fit),
+        (lgb.DaskLGBMRanker.predict, lgb.LGBMRanker.predict)
+
+    ]
+)
+def test_dask_methods_and_sklearn_equivalents_have_similar_signatures(methods):
+    dask_spec = inspect.getfullargspec(methods[0])
+    sklearn_spec = inspect.getfullargspec(methods[1])
+
+    # checks for a sublist in sklearn's kwargs which matches dask's
+    if dask_spec.args is not None and sklearn_spec.args is not None:
+        assert any(sklearn_spec.args[i: i + len(dask_spec.args)] == dask_spec.args for i in range(len(sklearn_spec.args) - len(dask_spec.args) + 1)) is True
+
+    if dask_spec.defaults is not None and sklearn_spec.defaults is not None:
+        assert any(sklearn_spec.defaults[i: i + len(dask_spec.defaults)] == dask_spec.defaults for i in range(len(sklearn_spec.defaults) - len(dask_spec.defaults) + 1)) is True
```

Comment:
Thanks very much for this! I think this code could be simplified a lot, and needs some additional comparison.

You can remove the `if dask_spec.args is not None`... blocks. All the methods being considered will have a non-empty set of arguments.

I don't think this list comprehension is necessary. Something like this should work:

```python
num_dask_args = len(dask_spec.args)
assert dask_spec.args == sklearn_spec.args[:num_dask_args]
assert dask_spec.defaults == sklearn_spec.defaults[:num_dask_args]
```

Basically, you can rely on the fact that the built-in `==` method for list objects will take care of most of what you've written out by hand.

<hr>

Can you please also include checks on all the other characteristics in the spec? https://github.com/microsoft/LightGBM/blob/fc2da2ec84e2dd9aa06421edcb5a625808eb546e/tests/python_package_test/test_dask.py#L819-L822

* `varargs`
* `varkw`
* `kwonlyargs`
* `kwonlydefaults`

If you need help inspecting these characteristics, you can use the `_asdict()` method on the results of `inspect.getfullargspec()`, like this:

```python
inspect.getfullargspec(methods[1])._asdict()
```

That produces a nice output like this

```text
OrderedDict([('args',
              ['self',
               'X',
               'y',
               'sample_weight',
               'init_score',
               'eval_set',
               'eval_names',
               'eval_sample_weight',
               'eval_init_score',
               'eval_metric',
               'early_stopping_rounds',
               'verbose',
               'feature_name',
               'categorical_feature',
               'callbacks',
               'init_model']),
             ('varargs', None),
             ('varkw', None),
             ('defaults',
              (None,
               None,
               None,
               None,
               None,
               None,
               None,
               None,
               True,
               'auto',
               'auto',
               None,
               None)),
             ('kwonlyargs', []),
             ('kwonlydefaults', None),
             ('annotations', {})])
```

</hr>

I did some of my own research while reviewing this, and I found a confusing behavior. It seem defaults for these specs are not being parsed the way I'd expect. For example, `lgb.DaskLGBMClassifier.fit()` takes 4 arguments. Its `.args` from `inspect.getfullargspec()` correctly has 4 items, but its `.defaults` only has 1!

```python
import inspect
import lightgbm as lgb
dask_spec = inspect.getfullargspec(lgb.DaskLGBMClassifier.fit)
sklearn_spec = inspect.getfullargspec(lgb.LGBMClassifier.fit)

print(dask_spec.args)
# ['self', 'X', 'y', 'sample_weight']

print(dask_spec.defaults)
# (None,)

print(sklearn_spec.args)
# ['self', 'X', 'y', 'sample_weight', 'init_score', 'eval_set', 'eval_names', 'eval_sample_weight', 'eval_init_score', 'eval_metric', 'early_stopping_rounds', 'verbose', 'feature_name', 'categorical_feature', 'callbacks', 'init_model']

print(sklearn_spec.defaults)
# (None, None, None, None, None, None, None, None, True, 'auto', 'auto', None, None)
```

After some investigation, I found that the `.defaults` list is an ordered list of defaults **only for the arguments that have defaults**. This makes it really hard to compare two different functions.

For the comparison, could you try looping over the values of `dask_spec.args`, and then checking that the defaults are the same as the equivalent sklearn default by using `inspect.signature()`?

For example:

```python
dask_params = inspect.signature(lgb.DaskLGBMClassifier.fit).parameters
sklearn_params = inspect.signature(lgb.LGBMClassifier.fit).parameters

print(dask_params['X'].default)
print(sklearn_params['X'].default)

dask_params['X'].default == sklearn_params['X'].default
```

Reply:
Thanks for the feedback! The only reason i put `if dask_spec.args is not None` , was that it was failing the code checks,
because of the error `TypeError: object of type 'NoneType' has no len()` , so I figured I had to put that.
 
As for the defaults, yeah it makes sense, because the `.defaults` list is containing only those args that have defaults, my approach wouldn't work as there could be arguments in between that don't have defaults. 

Will make the necessary changes.

Reply:
Please let me know if you hit that NoneType error again. I'd be surprised to learn that `.args` for any of the `fit()` or `predict()` methods returns `None`.

Thanks for contributing!